### PR TITLE
Feature: Schema overriding

### PR DESCRIPTION
**Pull request by fulghum**
Allows customers to specify a commit, branch, or tag in the `@@dolt_schema_override_commit` session variable and have all table's data mapped to the schema from that commit, branch, or tag, when queried. 

### Example
As a simple example, consider a database with a `main` branch that has added the new column `birthday` to a table, and an `olderBranch` branch with a table that has not been updated with that schema change. Customers cannot use the same queries from the `main` branch to query the data on the `olderBranch` because of the schema difference. Setting a schema override allows the customer to map the table schemas on the `olderBranch` branch to the same schema as on the `main` branch. This can be useful when you want to run queries on older data, but don't want to rewrite your queries for older schemas. 

```sql
CALL dolt_checkout(‘olderBranch’);

SELECT name, birthday from people;
column "birthday" could not be found in any table in scope

SET @@dolt_schema_override_commit = ‘main’;
SELECT name, birthday from people;
+-------+----------+
| name  | birthday |
+-------+----------+
| Sammy | NULL     |
+-------+----------+
```

### Limitations
The first version of schema override support is subject to several limitations. [Please reach out to us](https://github.com/dolthub/dolt/issues/new) and let us know if you'd like any of these to be prioritized. 
- **Read-only** – when a schema override has been set, only read queries can be executed. Attempting to write data or execute DDL will result in an error about the database being read-only. 
- **System tables** – Dolt system tables currently do not honor schema overrides. 
- **Collation changes** – Collations affect how data is sorted when stored on disk. To map data from one collation to another collation requires extra processing to ensure the returned results are sorted according to the mapped collation. This extra processing is not supported yet, so collation changes will not appear when overriding a schema. 
- **Column defaults** – If the overridden schema has added new columns with column defaults, those column defaults do not currently get applied when that column is queried. Using the column default value, instead of `NULL` is a planned enhancement. 

[Design doc](https://docs.google.com/document/d/1Ox068KmXHfBDcAzUeark17EwZeqXMiZEy0EyGM0_Hn8/edit?usp=sharing)

Reference docs update: https://github.com/dolthub/docs/pull/2062

Fixes: https://github.com/dolthub/dolt/issues/5486

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>8f2eb2e</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>8f2eb2e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>1f3579d</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>1f3579d</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/enginetest/dolt_queries_schema_override.go [hunk 1]
```diff
@@ -0,0 +1,1034 @@
...
+			{
+				// Going back to @commit1 for data, but using @commit2 for schema
+				Query:           "SELECT c1 from t as of @commit1 where c1 > 'o';",
+				Expected:        []sql.Row{{"one"}},
+				ExpectedIndexes: []string{"c1_idx"},
+			},
+		},
+	},
+
+	// AS OF TEST CASES
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/enginetest/dolt_queries_schema_override.go:713]: Other forms you're missing that could use some coverage:

```
`database/branch`.`table`
`database/HEAD~`.`table`
`database/tag`.`table`
`database/commitHash`.`table`
```

**Review by zachmu — APPROVED** [review]
Overall this looks great, nice work.

There are probably things we haven't thought of but the tests look good enough for an initial release.

File: go/libraries/doltcore/sqle/history_table.go [hunk 2]
```diff
@@ -172,6 +172,12 @@ func (ht *HistoryTable) LookupPartitions(ctx *sql.Context, lookup sql.IndexLooku
 func NewHistoryTable(table *DoltTable, ddb *doltdb.DoltDB, head *doltdb.Commit) sql.Table {
 	cmItr := doltdb.CommitItrForRoots(ddb, head)
 
+	// System tables don't currently use overridden schemas, so if one is set on |table|,
+	// clear it out to make sure we use the correct schema that matches the data being used.
+	if table.overriddenSchema != nil {
```
[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/sqle/history_table.go:177]: Good call; added tests in d89cce9ae00fd786091f78c91a196c3cc9a1987e

File: go/libraries/doltcore/sqle/schema_override.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	tbl.OverrideSchema(overriddenSchema)
+	return nil
+}
+
+// getOverriddenSchemaValue returns a pointer to the string value of the Dolt schema override session variable. If the
+// variable is not set (i.e. NULL or empty string) then this function returns nil so that callers can check for any
+// set value by testing against nil.
+func getOverriddenSchemaValue(ctx *sql.Context) (*string, error) {
```
[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/sqle/schema_override.go:89]: Nah, I was experimenting with a `*string` to see if it made the calling code any cleaner, but it's not very standard in the code base and doesn't really make much of a difference in the calling code. I'll switch back to just `string` and keep this a little simpler.

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d89cce9</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d89cce9</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/schema_override.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+		return nil, nil
+	}
+
+	commitSpec, err := doltdb.NewCommitSpec(*overriddenSchemaValue)
+	if err != nil {
+		return nil, fmt.Errorf("invalid commit spec specified in %s: %s", dsess.DoltOverrideSchema, err.Error())
+	}
+
+	doltSession := dsess.DSessFromSess(ctx.Session)
+	headRef, err := doltSession.CWBHeadRef(ctx, db.Name())
```
[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/sqle/schema_override.go:131]: Ahh... that makes sense. Thanks for pointing that out!

File: go/libraries/doltcore/sqle/enginetest/dolt_queries_schema_override.go (hunk 1 above)
Comment (fulghum) [reply go/libraries/doltcore/sqle/enginetest/dolt_queries_schema_override.go:713]: Good call – tag revision and commit hash revision syntax triggered a panic on a bad type cast. Fixed and added tests.

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>fd7be9a</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>fd7be9a</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>95a5964</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>95a5964</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

File: go/libraries/doltcore/sqle/schema_override.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+// behavior. It would be nice to resolve the differences and standardize on how we convert rows between schemas.
+// The main differences are:
+//  1. The dolt_history_ system tables only maps columns by name and doesn't take into account tags. This
+//     implementation prefers mapping by column tags, but will fall back to column names if a column with a specified
+//     tag can't be found. This behavior is similar to what we do in the diff system tables. Related to this, the
+//     columns to include in the projection are also only specified by name in the dolt_history system tables, but
+//     here they need to be specified by tag and then fallback to column name matching if a tag isn't found.
+//  2. The dolt_history_ system tables will not map columns unless their types are exactly identical. This is too
+//     strict for schema override mapping, so this implementation attempts to convert column values to the target
+//     type. If a column value is not compatible with the mapped column type (e.g. a 10-char string in a varchar(5)
```
[the remark this answers is not in the document]
Comment (fulghum) [reply go/libraries/doltcore/sqle/schema_override.go:164]: I think that makes sense for char/varchar/text. I don't think there are many other type conversions where that makes sense though. Seems worth logging a warning for good measure, too, even though people probably rarely ever notice those. Added string truncation in 95a5964e04fe52ffbeac9d2de0dea1d7ea0af7c2.

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>794e30d</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>794e30d</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@fulghum __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d2c4af2</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d2c4af2</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**